book/src/proposals/booting-from-snapshot.md

```diff
@@ -0,0 +1,222 @@
+as it will be verified before being vote upon. Call the root of this snapshot
+`S`.
+
+2) We reject any snapshots `S` where `S` is less than the root of `L` and `S`
+is not in the list of roots in blocktree because this means `S` is for a
+different fork than the one this validiator last rooted. Thus it's critical for
+consistency in `replay_stage` that the order of events when setting a new root
+is:
+
+    1) Write the root to tower
+    2) Write the root to blocktree
+    3) Generate snapshot for that root
+
+3) On startup, the validator checks that the current root in tower exists
+in blocktree. If not (there was a crash between 2-1 and 2-2), then rewrite the
+root to blocktree.
+
+4) On startup, the validator boots from the snapshot `S`, then replays all
+descendant blocks of `S` that exist in this validator's ledger, building banks
+which are then stored in an output `BankForks`. This is done in
+`blocktree_processor.rs`. The root of this `BankForks` is set to `S`.
+
+5) On startup, the validator calculates the first slot `S_n` from which it is
+not locked out for voting. Every validator persists its tower state and
+must consult this state in order to boot safely and resume from a snapshot
+without being slashed. From this tower state and the ancestry information
+embedded in the snapshot, a validiator can derive which banks, are "safe" (See
+the `Determining Vote Safety From a Snapshot and Tower` section for more
+detals) to vote for.
+
+6) Periodically send canary transactions to the cluster using the validator's
+local recent blockhash.
+
+7) Wait for the following criteria. While waiting, set a new root every time
+2/3 of the cluster's stake roots a bank. This allows the validator to prune its
+state and also calculate leader schedules as it moves across epochs.
+
+   1) Observe a canary transaction in a bank that some threshold of trusted
+   validator stake has voted on. Call this trusted bank `T`.
+   
+   2) The current working bank has slot number `S_current` > `S_n`
+
+8) Start the normal voting process, voting for any slot that satisfies
+`S_current` > `S_n`
+
+
+## Determining `S_n` From a Snapshot and Tower
+
+Define the "safety" condition to be: 
+
+Given a `BankForks`, a validator can run some procedure to determine the first
+slot `S_n` that it can vote on without violating the lockouts of any vote in
+`L`. The procedure for this is:
+
+1) Find the earliest vote `V` in the tower for which `is_ancestor(V, S)` is not
+true.
+
+2) `V` and every vote after it needs to expire before the validator can vote on
+any descendant of `S`, so with all the votes `V_i` in tower where `V_i >=
+V`, `S_n = max(V_i + lockout(V_i))`. If no `V` exists, then `S_n = S`.
+
+
+## Achieving Safety
+
+Define the "Safety Criteria" to be: Given a `BankForks` any descendant of `S`,
+`S_d` that has a bank `B_d` that is present in `BankForks.banks`, and any slot
+`L_i` in tower, the validator is able to determine `is_ancestor(L_i, S_d)`.
```

Comment:
[nits] Preferably, in addition to formally defining `B_d` [moving up from there](https://github.com/solana-labs/solana/pull/6936/files#r354170535), it would be really nice we could break this rather long sentence into some. I had hard time to discern which qualify which... ;)

Reply:
How's this update look?